libncftp/samples/ncftpls/ncftpls.1
.TH ncftpls 1 "NcFTP Software" ncftpls
.SH NAME
ncftpls - Internet file transfer program for scripts
.SH "SYNOPSIS"
.PP
.B ncftpls
.RI [ "options" ]
.I "ftp://url.style/host/path/name/"
.\"-------
.SH "OPTIONS"
.\"-------
.SS
Command line flags:
.TP 8
\-1
Most basic format, one item per line.
.TP 8
\-l
Long list format.
.TP 8
.RI "-x -" "XX"
Additional
.I ls
flags to pass on to the server.
.TP 8
.RI "-u " "XX"
Use username
.I XX
instead of anonymous.
.TP 8
.RI "-p " "XX"
Use password
.I XX
with the username.
.TP 8
.RI "-P " "XX"
Use port number
.I XX
instead of the default FTP service port (21).
.TP 8
.RI "-d " "XX"
Use the file
.I XX
for debug logging.
.TP 8
.RI "-t " "XX"
Timeout after
.I XX
seconds.
.TP 8
-F
Use passive (PASV) data connections.
.TP 8
.RI "-r " "XX"
Redial a maximum of 
.I XX
times until connected to the remote FTP server.
.\"-------
.SH "DESCRIPTION"
.\"-------
.PP
The
purpose of
.I ncftpls
is to do remote directory listings using
the File Transfer Protocol without entering an interactive shell.
This lets you write shell scripts or other unattended
processes that can do FTP.
.PP
The default behavior is to print the directory listing in columnized
format (i.e. ls\ \-CF), but that is not very useful for scripting.
This example uses the
.B \-1
flag, to print one file per line:
.RS 4
.br
.sp
$ ncftpls -1 ftp://ftp.ncftp.com/pub/ncftp/
.RE
.PP
You can also do a remote "ls\ \-l", by using "ncftpls\ \-l".
If you want to try other flags, you have to use them with the
.B \-x
flag.
For example, if you wanted to do a remote "ls\ \-lrt", you could
do this:
.RS 4
.br
.sp
$ ncftpls -x "-lrt" ftp://ftp.ncftp.com/pub/ncftp/
.RE
.PP
By default the program tries to open the remote host
and login anonymously, but you can specify a username
and password information like you can with
.I ncftpget
or
.IR ncftpput "."
.\"-------
.SH "DIAGNOSTICS"
.\"-------
.PP
.I ncftpls
returns the following exit values:
.TP 8
0
Success.
.TP 8
1
Could not connect to remote host.
.TP 8
2
Could not connect to remote host - timed out.
.TP 8
3
Transfer failed.
.TP 8
4
Transfer failed - timed out.
.TP 8
5
Directory change failed.
.TP 8
6
Directory change failed - timed out.
.TP 8
7
Malformed URL.
.TP 8
8
Usage error.
.TP 8
9
Error in login configuration file.
.TP 8
10
Library initialization failed.
.TP 8
11
Session initialization failed.
.\"-------
.SH "AUTHOR"
.\"-------
.PP
Mike Gleason, NcFTP Software (http://www.ncftp.com).
.\"-------
.SH "SEE ALSO"
.\"-------
.PP
.IR ncftpput (1),
.IR ncftpget(1),
.IR ncftp (1),
.IR ftp (1),
.IR rcp (1),
.IR tftp (1).
.PP
.IR "LibNcFTP" " (http://www.ncftp.com/libncftp/)."


doc/CHANGELOG.txt
LibNcFTP Change Log:
===================

3.2.6, 2016-11-12

   + If a recursive download operation is also requested with delete mode,
     attempt to remove empty directories after all files have completed
     successfully.

   + To avoid problems on servers that penalize clients for trying to CWD
     with a non-directory, FTPRmdirRecursive now ensures sub-pathnames
     are directories before trying FTPChdir.

   + No longer trying to utime() after every single block on downloads,
     which could cause noticable performance degradation when the local
     filesystem was not local.

   + Changed behavior of resuming downloads where the timestamp wasn't
     preserved (because of the utime change, above). The new behavior is
     to resume the download when the local copy has a recent timestamp
     (less than a week).

   + You can now disable use of MFMT like you could similarly disable
     SITE UTIME.

   + You can now explicitly set hasAPPE to 0 to prevent it being used
     when hasREST is also 0.

   + The library now caches the most recent value of PWD and uses it
     to prevent redundant PWD commands.

   + Now able to use sendfile() for uploads, on Linux/FreeBSD/Mac.
     Progress reports work too, with a small performance penalty.


3.2.5, 2011-01-17

   + If a server's MLSx output returns both the "UNIX.mode" and the "perm"
     parameters, UNIX.mode will take precedence (Thanks, Jay Kulpinski).


3.2.4, 2010-04-07

   + Compatibility fixes for FreeBSD 8.

   + Support for local validation of DNSSEC when combined with libraries
     available from the DNSSEC Tools project (Thanks, Robert Story).

   + Directory listings now assume a line with no year implies that the
     current year should be used if the date is no more than 2 days into
     the future.  Previously, for UNIX ls compatibility, up to 6 months into
     the future was allowed.

   + Compatibility fixes for MinGW32 (Windows GCC).

   + Microsoft disabled recursive directory listings altogether when fixing
     KB975254 (Thanks, Andrew Coggeshall).  When entire directories are
     downloaded, instead of using "LIST -R" for one efficient listing of
     all files, the directories are manually traversed with one directory
     listing for each subdirectory.  This will improve reliability of
     recursive downloads, except for the case when their are circular
     symbolic links (which is why "LIST -R" had been favored, which the
     server can easily handle).

   + Compatibility fixes for Mac OS X for building from source code
     (Thanks, Mathieu Rene)



3.2.3, 2009-07-24

   + Using MFMT command to set remote timestamps, if available, in
     preference to SITE UTIME or MDTM.

   + New configure flag, --enable-ssp, will turn on stack smashing
     protection (if available by compiler) by adding to CFLAGS.

   + Host, user, and pass fields have been increased in size.

   + New debugTimestamping boolean field has been added, which will
     cause the date and time to be printed with each call to the
     logging functions.


3.2.2.2, 2008-10-29

   + Large File Support has been updated internally to use newer APIs,
     rather than the transitional API.

   + You can now specify an IP address for binding the local end of
     sockets.  Useful if you need connections to originate from a
     specific IP on the system.

   + Bug fixed where binding an ephemeral port number in a specified
     range (rare) may not have used all available retries for binding.

   + Local writes now "retry" when the write returns less than the full
     number of bytes specified, and will call write() again to write
     out the remaining bytes.  This bug is very rare since local write
     almost always write out the full number of bytes.  This bug may
     have been limited to a particular type of filesystem.

   + Using a higher resolution timer on Windows, so that the timer
     no longer wraps over after a month.

   + The minimum version of Windows that the source code will now
     compile on is Windows 2000, although with a little tweaking you
     could compile for Windows NT 4 again.

   + Compatibility fix for Mac OS X when compiling from source to build
     universal binaries (Thanks, Toshi NAGATA)

   + Compatibility fix for FreeBSD, where a connection attempt may
     have failed with "Operation now in progress."


3.2.2, 2008-08-08

  + PASV mode will now be retried, in case of a transient failure.
    (Thanks, Steven Frank).

  + The directory listing parser now can recognize dates in the format
    of YYYY-mm-dd HH:MM[:SS].

  + Fixed feature detection for WS_FTP servers.

  + Workaround bug in REST on WS_FTP 6.0 (Thanks, Dan Nelson).

  + Fixed a problem where doing a get on a remote file that did not exist,
    but a file by the same name existed locally, would err out but truncate
    the local file anyway.

  + On Mac OS X, the configure script will now try to build universal
    binaries by default.  If you don't want that, pass --disable-universal
    to configure.

  + The default connect timeout has been reduced from 30 to 10 seconds to 
    reflect the increase in the typical user's network speed.

  + Visual Studio 2008 solution files are now provided for building the
    source on Windows.


3.2.1, 2007-07-29

  + Now a little more tolerant of FTP servers who send their internal network
    IP addresses in their PASV response.  When this is detected, it will be
    ignored and the data connection will connect to the same IP that is used
    for the control connection.

  + Fix problem where an empty directory could cause a recursive download
    to abort prematurely.
  
  + Resuming of uploads can now work for servers that do not support
    REST + STOR.  The APPE command will be used instead.

  + A few memory leaks plugged.

  + Fixes for FreeBSD 6 for compatibility with /usr/bin/make.


3.2.0, 2006-08-05

  + Fix read overflow in FTW (Thanks, Carl MAILLOUX, Craig Ball, and others).


libncftp/samples/ncftpls/ncftpls.c
/* ncftpls.c
 *
 * A non-interactive utility to list directories on a remote FTP server.
 * Very useful in shell scripts!
 *
 * NOTE:  This is for demonstration only -- an up-to-date version of this
 * program comes with the NcFTP Client 3.0 distribution.  This version
 * will most likely have a subset of features and fixes.
 */

#define VERSION "1.6.0"

#if (defined(WIN32) || defined(_WINDOWS)) && !defined(__CYGWIN__)
#	define _CRT_SECURE_NO_WARNINGS 1
#endif

#ifdef HAVE_CONFIG_H
#	include <config.h>
#endif

#include <ncftp.h>				/* Library header. */
#include <Strn.h>				/* Library header. */

#if (defined(WIN32) || defined(_WINDOWS)) && !defined(__CYGWIN__)
#	include "..\ncftpget\gpshare.h"
#else
#	include "../ncftpget/gpshare.h"
#endif

FTPLibraryInfo li;
FTPConnectionInfo fi;

extern int gFirewallType;
extern char gFirewallHost[64];
extern char gFirewallUser[32];
extern char gFirewallPass[32];
extern unsigned int gFirewallPort;

static void
Usage(void)
{
	FILE *fp;

	fp = OpenPager();
	(void) fprintf(fp, "NcFTPLs %s.\n\n", VERSION);
	(void) fprintf(fp, "*** NOTE: This version is for demonstration only -- get the latest and most\nfeature complete version from the NcFTP Client package. ***\n\n");
	(void) fprintf(fp, "Usages:\n");
	(void) fprintf(fp, "  ncftpls [FTP flags] [-x \"ls flags\"] ftp://url.style.host/path/name/\n");
	(void) fprintf(fp, "\nls Flags:\n\
  -1     Most basic format, one item per line.\n\
  -l     Long list format.\n\
  -x XX  Other flags to pass on to the remote server.\n");
	(void) fprintf(fp, "\nFTP Flags:\n\
  -u XX  Use username XX instead of anonymous.\n\
  -p XX  Use password XX with the username.\n\
  -P XX  Use port number XX instead of the default FTP service port (21).\n\
  -d XX  Use the file XX for debug logging.\n");
	(void) fprintf(fp, "\
  -t XX  Timeout after XX seconds.\n\
  -f XX  Read the file XX for user and password information.\n\
  -F     Use passive (PASV) data connections.\n\
  -r XX  Redial XX times until connected.\n");
	(void) fprintf(fp, "\nExamples:\n\
  ncftpls ftp://ftp.wustl.edu/pub/\n\
  ncftpls -1 ftp://ftp.wustl.edu/pub/\n\
  ncftpls -x \"-lrt\" ftp://ftp.wustl.edu/pub/\n");

	(void) fprintf(fp, "\nLibrary version: %s.\n", gLibNcFTPVersion + 5);
	(void) fprintf(fp, "\nThis is a freeware program by Mike Gleason (http://www.NcFTP.com/contact/).\n");
	(void) fprintf(fp, "This was built using LibNcFTP (http://www.ncftp.com/libncftp).\n");

	ClosePager(fp);
	exit(kExitUsage);
}	/* Usage */





static void
Abort(int sigNum)
{
	signal(sigNum, Abort);

	/* Hopefully the I/O operation in progress
	 * will complete, and we'll abort before
	 * it starts a new block.
	 */
	fi.cancelXfer++;

	/* If the user appears to be getting impatient,
	 * restore the default signal handler so the
	 * next ^C abends the program.
	 */
	if (fi.cancelXfer >= 2)
		signal(sigNum, SIG_DFL);
}	/* Abort */




static void
SetLsFlags(char *dst, size_t dsize, int *longMode, const char *src)
{
	char *dlim = dst + dsize - 1;
	int i, c;

	for (i=0;;) {
		c = *src++;
		if (c == '\0')
			break;
		if (c == 'l') {
			*longMode = 1;
		} else if (c == '1') {
			*longMode = 0;
		} else if (c != '-') {
			if (c == 'C') {
				*longMode = 0;
			}
			if (i == 0) {
				if (dst < dlim)
					*dst++ = '-';
			} 
			i++;
			if (dst < dlim)
				*dst++ = (char) c;
		}
	}
	*dst = '\0';
}	/* SetLsFlags */





main_void_return_t
main(int argc, char **argv)
{
	int result, c;
	FTPConnectionInfo savedfi;
	FTPConnectionInfo startfi;
	ExitStatus es;
	char url[256];
	char urlfile[128];
	char rootcwd[256];
	int longMode = 0;
	int i;
	char lsflag[32] = "";
	FTPLineList cdlist;
	FTPLinePtr lp;
	int rc;
	int ndirs;
	GetoptInfo opt;

	InitWinsock();
	result = FTPInitLibrary(&li);
	if (result < 0) {
		(void) fprintf(stderr, "ncftpls: init library error %d (%s).\n", result, FTPStrError(result));
		DisposeWinsock();
		exit(kExitInitLibraryFailed);
	}
	result = FTPInitConnectionInfo(&li, &fi, kDefaultFTPBufSize);
	if (result < 0) {
		(void) fprintf(stderr, "ncftpls: init connection info error %d (%s).\n", result, FTPStrError(result));
		DisposeWinsock();
		exit(kExitInitConnInfoFailed);
	}

	fi.debugLog = NULL;
	fi.errLog = stderr;
	fi.xferTimeout = 60 * 60;
	fi.connTimeout = 30;
	fi.ctrlTimeout = 135;
	(void) STRNCPY(fi.user, "anonymous");
	fi.host[0] = '\0';
	urlfile[0] = '\0';
	InitLineList(&cdlist);
	SetLsFlags(lsflag, sizeof(lsflag), &longMode, "-CF");
	es = kExitSuccess;

	GetoptReset(&opt);
	while ((c = Getopt(&opt, argc, argv, "1lx:P:u:p:e:d:t:r:f:F")) > 0) switch(c) {
		case 'P':
			fi.port = atoi(opt.arg);	
			break;
		case 'u':
			(void) STRNCPY(fi.user, opt.arg);
			break;
		case 'p':
			(void) STRNCPY(fi.pass, opt.arg);	/* Don't recommend doing this! */
			break;
		case 'e':
			if (strcmp(opt.arg, "stdout") == 0)
				fi.errLog = stdout;
			else if (opt.arg[0] == '-')
				fi.errLog = stdout;
			else if (strcmp(opt.arg, "stderr") == 0)
				fi.errLog = stderr;
			else
				fi.errLog = fopen(opt.arg, "a");
			break;
		case 'd':
			if (strcmp(opt.arg, "stdout") == 0)
				fi.debugLog = stdout;
			else if (opt.arg[0] == '-')
				fi.debugLog = stdout;
			else if (strcmp(opt.arg, "stderr") == 0)
				fi.debugLog = stderr;
			else
				fi.debugLog = fopen(opt.arg, "a");
			break;
		case 't':
			SetTimeouts(&fi, opt.arg);
			break;
		case 'r':
			SetRedial(&fi, opt.arg);
			break;
		case 'f':
			ReadConfigFile(opt.arg, &fi);
			break;
		case 'F':
			if (fi.dataPortMode == kPassiveMode)
				fi.dataPortMode = kSendPortMode;
			else
				fi.dataPortMode = kPassiveMode;
			break;
		case 'l':
			SetLsFlags(lsflag, sizeof(lsflag), &longMode, "-l");
			break;
		case '1':
			SetLsFlags(lsflag, sizeof(lsflag), &longMode, "-1");
			break;
		case 'x':
			SetLsFlags(lsflag, sizeof(lsflag), &longMode, opt.arg);
			break;
		default:
			Usage();
	}
	if (opt.ind > argc - 1)
		Usage();

	InitOurDirectory();
	LoadFirewallPrefs();

	startfi = fi;
	memset(&savedfi, 0, sizeof(savedfi));
	ndirs = argc - opt.ind;
	for (i=opt.ind; i<argc; i++) {
		fi = startfi;
		(void) STRNCPY(url, argv[i]);
		rc = FTPDecodeURL(&fi, url, &cdlist, urlfile, sizeof(urlfile), (int *) 0, NULL);
		(void) STRNCPY(url, argv[i]);
		if (rc == kMalformedURL) {
			(void) fprintf(stderr, "Malformed URL: %s\n", url);
			DisposeWinsock();
			exit(kExitMalformedURL);
		} else if (rc == kNotURL) {
			(void) fprintf(stderr, "Not a URL: %s\n", url);
			DisposeWinsock();
			exit(kExitMalformedURL);
		} else if (urlfile[0] != '\0') {
			/* It not obviously a directory, and they didn't say -R. */
			(void) fprintf(stderr, "Not a directory URL: %s\n", url);
			DisposeWinsock();
			exit(kExitMalformedURL);
		}
		
		if ((strcmp(fi.host, savedfi.host) == 0) && (strcmp(fi.user, savedfi.user) == 0)) {
			fi = savedfi;
			
			/* This host is currently open, so keep using it. */
			if (FTPChdir(&fi, rootcwd) < 0) {
				(void) fprintf(stderr, "ncftpls: cannot chdir to %s: %s.\n", rootcwd, FTPStrError(fi.errNo));
				es = kExitChdirFailed;
				DisposeWinsock();
				exit((int) es);
			}
		} else {
			if (savedfi.connected != 0) {
				(void) FTPCloseHost(&savedfi);
			}
			memset(&savedfi, 0, sizeof(savedfi));

			if ((strcmp(fi.user, "anonymous") != 0) && (strcmp(fi.user, "ftp") != 0) && (fi.pass[0] == '\0'))
				(void) GetPass("Password: ", fi.pass, sizeof(fi.pass));
			
			if (MayUseFirewall(fi.host) != 0) {
				fi.firewallType = gFirewallType; 
				(void) STRNCPY(fi.firewallHost, gFirewallHost);
				(void) STRNCPY(fi.firewallUser, gFirewallUser);
				(void) STRNCPY(fi.firewallPass, gFirewallPass);
				fi.firewallPort = gFirewallPort;
			}
			
			es = kExitOpenTimedOut;
			if ((result = FTPOpenHost(&fi)) < 0) {
				(void) fprintf(stderr, "ncftpls: cannot open %s: %s.\n", fi.host, FTPStrError(result));
				es = kExitOpenFailed;
				DisposeWinsock();
				exit((int) es);
			}
			
			if (FTPGetCWD(&fi, rootcwd, sizeof(rootcwd)) < 0) {
				(void) fprintf(stderr, "ncftpls: cannot pwd: %s.\n", FTPStrError(fi.errNo));
				es = kExitChdirFailed;
				DisposeWinsock();
				exit((int) es);
			}			
		}
		
		es = kExitChdirTimedOut;
		for (lp = cdlist.first; lp != NULL; lp = lp->next) {
			if (FTPChdir(&fi, lp->line) != 0) {
				(void) fprintf(stderr, "ncftpls: cannot chdir to %s: %s.\n", lp->line, FTPStrError(fi.errNo));
				es = kExitChdirFailed;
				DisposeWinsock();
				exit((int) es);
			}
		}
		
		if (ndirs > 1) {
			fprintf(stdout, "%s%s\n\n",
				(i > opt.ind) ? "\n\n\n" : "", url);
		}
		fflush(stdout);
		
		es = kExitXferTimedOut;
		(void) signal(SIGINT, Abort);
		if (FTPList(&fi, STDOUT_FILENO, longMode, lsflag) < 0) {
			(void) fprintf(stderr, "ncftpls: directory listing error: %s.\n", FTPStrError(fi.errNo));
			es = kExitXferFailed;
		} else {
			es = kExitSuccess;
			savedfi = fi;
		}
		(void) signal(SIGINT, SIG_DFL);
	}
	
	(void) FTPCloseHost(&fi);

	DisposeWinsock();
	exit((int) es);
}	/* main */
